engine/src/main/java/io/zeebe/engine/processor/workflow/job/JobErrorThrownProcessor.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.0. You may not use this file
+ * except in compliance with the Zeebe Community License 1.0.
+ */
+package io.zeebe.engine.processor.workflow.job;
+
+import io.zeebe.engine.processor.KeyGenerator;
+import io.zeebe.engine.processor.TypedRecord;
+import io.zeebe.engine.processor.TypedRecordProcessor;
+import io.zeebe.engine.processor.TypedResponseWriter;
+import io.zeebe.engine.processor.TypedStreamWriter;
+import io.zeebe.engine.processor.workflow.deployment.model.element.ExecutableActivity;
+import io.zeebe.engine.processor.workflow.deployment.model.element.ExecutableBoundaryEvent;
+import io.zeebe.engine.processor.workflow.deployment.model.element.ExecutableWorkflow;
+import io.zeebe.engine.state.deployment.WorkflowState;
+import io.zeebe.engine.state.instance.ElementInstance;
+import io.zeebe.engine.state.instance.ElementInstanceState;
+import io.zeebe.engine.state.instance.EventScopeInstanceState;
+import io.zeebe.protocol.impl.record.value.job.JobRecord;
+import io.zeebe.protocol.record.intent.WorkflowInstanceIntent;
+import org.agrona.DirectBuffer;
+import org.agrona.concurrent.UnsafeBuffer;
+
+public class JobErrorThrownProcessor implements TypedRecordProcessor<JobRecord> {
+
+  private static final DirectBuffer NO_VARIABLES = new UnsafeBuffer();
+
+  private final WorkflowState workflowState;
+  private final KeyGenerator keyGenerator;
+
+  private final FoundCatchEvent foundCatchEvent = new FoundCatchEvent();
+  private final ElementInstanceState elementInstanceState;
+  private final EventScopeInstanceState eventScopeInstanceState;
+
+  public JobErrorThrownProcessor(
+      final WorkflowState workflowState, final KeyGenerator keyGenerator) {
+    this.keyGenerator = keyGenerator;
+    this.workflowState = workflowState;
+    elementInstanceState = workflowState.getElementInstanceState();
+    eventScopeInstanceState = workflowState.getEventScopeInstanceState();
+  }
+
+  @Override
+  public void processRecord(
+      final TypedRecord<JobRecord> record,
+      final TypedResponseWriter responseWriter,
+      final TypedStreamWriter streamWriter) {
+
+    final var job = record.getValue();
+    final var serviceTaskInstanceKey = job.getElementInstanceKey();
+    final var serviceTaskInstance = elementInstanceState.getInstance(serviceTaskInstanceKey);
+
+    if (serviceTaskInstance != null && serviceTaskInstance.isActive()) {
+
+      final var errorCode = job.getErrorCodeBuffer();
+      final var workflow = getWorkflow(job.getWorkflowKey());
+
+      final var foundCatchEvent = findCatchEvent(workflow, serviceTaskInstance, errorCode);
```

Comment:
Personally, I don't like this pattern of setting a field in the function but then returning it anyway. Usually the variable then shadows the field and then it forces me to see if they're actually the same object and if not then the variable name should be changed. I get that it removes the allocation from the hot path but, as you said in a comment bellow, we're assuming errors happen rarely so it wouldn't be a big deal to have it in `findCatchEvent`. These are just my 2 cents and it's something we do frequently, so I'm fine with keeping it. It makes me wonder however, if it makes sense to keep the "Local variables should not shadow class fields" code smell on SonarCloud.

Reply:
Agree. I also don't like this style but I'm not sure how to make it better - if it should be allocation free. I renamed the field to not shadow the local variable.